### ISSUE TITLE
Problem of optimizing policy

### ISSUE THREAD

**Issue by zsano1** [shariqiqbal2810/MAAC#20]
Hi Shariq, first thank you for your code! And it works well. But when optimizing policy, shouldn't it be `probs * (-pol_target)`? Why we use `log_pi` here?
https://github.com/shariqiqbal2810/MAAC/blob/bd263afce709795293964badd16655b5747b9056/algorithms/attention_sac.py#L150

**Comment by shariqiqbal2810** [comment]
Hi,

This is the log derivative trick for estimating the gradient of the expected returns. [Here](http://www.1-4-5.net/~dmm/ml/log_derivative_trick.pdf) is a good explanation.

**Comment by zsano1** [comment]
Hi, thanks for your reply! But `log_pi` here is the log of one selected action right? In my opinion, we need `torch.log(probs)`, which is the log of policy distribution.

**Comment by shariqiqbal2810** [comment]
The purpose of the log is to account for sampling from the policy distribution (which estimates an expectation), so we only need the `log_pi` of the sampled action. If we wanted to use the probabilities of each action, we would also need all of their Q-values, and the method would resemble [Mean Actor-Critic](https://arxiv.org/pdf/1709.00503.pdf)

**Comment by zsano1** [comment]
Oh yes, you are right. I was assuming your `q` here is all actions' values. Thanks for your explanation.
